alica_engine/include/engine/BasicBehaviour.h

```diff
@@ -37,34 +37,34 @@ class BasicBehaviour
     virtual ~BasicBehaviour(){};
     virtual void run(void* msg) = 0;
 
-    bool isRunningInContext(const RunningPlan* rp) const;
+    bool isRunningInContext(const RunningPlan* rp) const { return isExecutingInContext() && rp == _context.load(); };
     void setEngine(AlicaEngine* engine) { _engine = engine; }
     const std::string& getName() const { return _name; }
 
-    void setBehaviour(const Behaviour* beh);
-    void setConfiguration(const Configuration* conf);
+    void setBehaviour(const Behaviour* beh) { _behaviour = beh; };
+    void setConfiguration(const Configuration* conf) { _configuration = conf; };
 
     const VariableGrp& getVariables() const { return _behaviour->getVariables(); }
     const Variable* getVariable(const std::string& name) const { return _behaviour->getVariable(name); };
 
     bool stop();
-    bool start();
-    void terminate();
+    bool start(RunningPlan* rp);
+    void terminate() { stop(); };
 
     void setDelayedStart(int32_t msDelayedStart) { _msDelayedStart = AlicaTime::milliseconds(msDelayedStart); }
 
     void setInterval(int32_t msInterval) { _msInterval = AlicaTime::milliseconds(msInterval); }
 
-    ThreadSafePlanInterface getPlanContext() const { return ThreadSafePlanInterface(isBehaviourStarted() ? _context : nullptr); }
-    void setRunningPlan(RunningPlan* rp) { _context = rp; }
+    ThreadSafePlanInterface getPlanContext() const { return ThreadSafePlanInterface(isExecutingInContext() ? _context.load() : nullptr); }
```

Comment:
It does not seem possible to guarantee that the pointer returned by getPlanContext() is safe to access in the behaviour unless a shared_ptr is used to ensure that the RunningPlan is not destroyed by the alica main engine thread. Even using a mutex here would not help